File: android/core/piyodeck/src/main/kotlin/app/piyokey/core/piyodeck/PiyoDeckZip.kt
package app.piyokey.core.piyodeck

import java.nio.charset.CodingErrorAction
import java.nio.charset.StandardCharsets
import java.util.zip.CRC32

internal data class PiyoDeckZipEntry(val name: String, val data: ByteArray)

internal object PiyoDeckZip {
  private const val LOCAL_FILE_HEADER_SIGNATURE: Long = 0x0403_4B50L
  private const val CENTRAL_DIRECTORY_SIGNATURE: Long = 0x0201_4B50L
  private const val END_OF_CENTRAL_DIRECTORY_SIGNATURE: Long = 0x0605_4B50L
  private const val UTF8_FLAG: Int = 0x0800
  private const val STORE_METHOD: Int = 0
  private const val ZIP_VERSION: Int = 20
  private const val DOS_TIME: Int = 0
  private const val DOS_DATE: Int = 0x0021
  private val requiredEntryNames = setOf("manifest.json", "deck.json")

  fun read(archive: ByteArray): Map<String, ByteArray> {
    if (archive.size > PiyoDeckPackageLimits.MAXIMUM_PACKAGE_BYTES) {
      throw PiyoDeckImportException.PackageTooLarge(
        archive.size,
        PiyoDeckPackageLimits.MAXIMUM_PACKAGE_BYTES,
      )
    }
    val endRecordSize = 22
    if (archive.size < endRecordSize) malformed("missing end-of-central-directory record")
    val endOffset = archive.size - endRecordSize
    if (archive.uint32(endOffset) != END_OF_CENTRAL_DIRECTORY_SIGNATURE) {
      malformed("archive comments, trailing data, or a missing end record")
    }

    val diskNumber = archive.uint16(endOffset + 4)
    val centralDirectoryDisk = archive.uint16(endOffset + 6)
    val entriesOnDisk = archive.uint16(endOffset + 8)
    val totalEntries = archive.uint16(endOffset + 10)
    val centralDirectorySize = archive.uint32(endOffset + 12)
    val centralDirectoryOffset = archive.uint32(endOffset + 16)
    val archiveCommentLength = archive.uint16(endOffset + 20)

    if (diskNumber == 0xFFFF || centralDirectoryDisk == 0xFFFF ||
      entriesOnDisk == 0xFFFF || totalEntries == 0xFFFF ||
      centralDirectorySize == 0xFFFF_FFFFL || centralDirectoryOffset == 0xFFFF_FFFFL
    ) {
      unsupported("ZIP64")
    }
    if (diskNumber != 0 || centralDirectoryDisk != 0) unsupported("split archive")
    if (archiveCommentLength != 0) unsupported("archive comment")
    if (entriesOnDisk != totalEntries) unsupported("multi-disk central directory")
    if (totalEntries != 2) throw PiyoDeckImportException.InvalidEntrySet(emptyList())
    if (centralDirectoryOffset > endOffset.toLong() ||
      centralDirectorySize != endOffset.toLong() - centralDirectoryOffset
    ) {
      malformed("invalid central-directory bounds")
    }

    val centralStart = centralDirectoryOffset.toInt()
    var cursor = centralStart
    val records = mutableListOf<CentralRecord>()
    repeat(totalEntries) {
      ensureRange(cursor, 46, endOffset)
      if (archive.uint32(cursor) != CENTRAL_DIRECTORY_SIGNATURE) {
        malformed("invalid central-directory entry")
      }
      val versionMadeBy = archive.uint16(cursor + 4)
      val versionNeeded = archive.uint16(cursor + 6)
      val flags = archive.uint16(cursor + 8)
      val method = archive.uint16(cursor + 10)
      val crc32 = archive.uint32(cursor + 16)
      val compressedSize = archive.uint32(cursor + 20)
      val uncompressedSize = archive.uint32(cursor + 24)
      val nameLength = archive.uint16(cursor + 28)
      val extraLength = archive.uint16(cursor + 30)
      val commentLength = archive.uint16(cursor + 32)
      val diskStart = archive.uint16(cursor + 34)
      val externalAttributes = archive.uint32(cursor + 38)
      val localHeaderOffset = archive.uint32(cursor + 42)

      if (versionNeeded > ZIP_VERSION) unsupported("ZIP version $versionNeeded")
      if (compressedSize == 0xFFFF_FFFFL || uncompressedSize == 0xFFFF_FFFFL ||
        localHeaderOffset == 0xFFFF_FFFFL
      ) {
        unsupported("ZIP64")
      }
      if (flags and UTF8_FLAG.inv() != 0) {
        unsupported(if (flags and 0x0001 != 0) "encryption" else "general-purpose flag 0x${flags.toString(16)}")
      }
      if (method != STORE_METHOD) unsupported("compression method $method")
      if (compressedSize != uncompressedSize) malformed("STORE entry sizes differ")
      if (extraLength != 0) unsupported("entry extra field")
      if (commentLength != 0) unsupported("entry comment")
      if (diskStart != 0) unsupported("split archive entry")

      val hostSystem = versionMadeBy ushr 8
      val unixFileType = ((externalAttributes ushr 16) and 0xF000).toInt()
      if (hostSystem == 3 || hostSystem == 19) {
        when (unixFileType) {
          0xA000 -> unsupported("symbolic link")
          0x4000 -> unsupported("directory entry")
          0, 0x8000 -> Unit
          else -> unsupported("non-regular entry")
        }
      }
      if (externalAttributes and 0x10L != 0L) unsupported("directory entry")

      val nameOffset = checkedEnd(cursor, 46, endOffset)
      val nameEnd = checkedEnd(nameOffset, nameLength, endOffset)
      val nameData = archive.copyOfRange(nameOffset, nameEnd)
      val name = decodeEntryName(nameData)
      records += CentralRecord(
        name = name,
        nameData = nameData,
        versionNeeded = versionNeeded,
        flags = flags,
        method = method,
        crc32 = crc32,
        compressedSize = compressedSize.toInt(),
        uncompressedSize = uncompressedSize.toInt(),
        localHeaderOffset = localHeaderOffset,
      )
      cursor = nameEnd
    }
    if (cursor != endOffset) malformed("unused central-directory bytes")

    val names = records.map { it.name }
    if (names.toSet() != requiredEntryNames || names.toSet().size != names.size) {
      throw PiyoDeckImportException.InvalidEntrySet(names)
    }

    val payloads = linkedMapOf<String, ByteArray>()
    val occupiedRanges = mutableListOf<IntRange>()
    records.forEach { record ->
      val maximum = if (record.name == "manifest.json") {
        PiyoDeckPackageLimits.MAXIMUM_MANIFEST_BYTES
      } else {
        PiyoDeckPackageLimits.MAXIMUM_DECK_BYTES
      }
      if (record.uncompressedSize > maximum) {
        throw PiyoDeckImportException.EntryTooLarge(record.name, record.uncompressedSize, maximum)
      }
      if (record.localHeaderOffset > Int.MAX_VALUE.toLong()) malformed("record exceeds archive bounds")
      val offset = record.localHeaderOffset.toInt()
      ensureRange(offset, 30, centralStart)
      if (archive.uint32(offset) != LOCAL_FILE_HEADER_SIGNATURE) {
        malformed("missing local header for ${record.name}")
      }
      val localVersionNeeded = archive.uint16(offset + 4)
      val localFlags = archive.uint16(offset + 6)
      val localMethod = archive.uint16(offset + 8)
      val localCrc32 = archive.uint32(offset + 14)
      val localCompressedSize = archive.uint32(offset + 18)
      val localUncompressedSize = archive.uint32(offset + 22)
      val localNameLength = archive.uint16(offset + 26)
      val localExtraLength = archive.uint16(offset + 28)
      if (localVersionNeeded > ZIP_VERSION) unsupported("ZIP version $localVersionNeeded")
      if (localFlags and UTF8_FLAG.inv() != 0) {
        unsupported(
          if (localFlags and 0x0001 != 0) {
            "encryption"
          } else {
            "general-purpose flag 0x${localFlags.toString(16)}"
          },
        )
      }
      if (localMethod != STORE_METHOD) unsupported("compression method $localMethod")
      if (localCompressedSize == 0xFFFF_FFFFL || localUncompressedSize == 0xFFFF_FFFFL) {
        unsupported("ZIP64")
      }
      if (localCompressedSize != localUncompressedSize) {
        malformed("STORE entry sizes differ for ${record.name}")
      }
      if (localVersionNeeded != record.versionNeeded ||
        localFlags != record.flags ||
        localMethod != record.method ||
        localCrc32 != record.crc32 ||
        localCompressedSize != record.compressedSize.toLong() ||
        localUncompressedSize != record.uncompressedSize.toLong()
      ) {
        malformed("local and central headers differ for ${record.name}")
      }
      if (localExtraLength != 0) unsupported("local entry extra field")
      val localNameOffset = checkedEnd(offset, 30, centralStart)
      val localNameEnd = checkedEnd(localNameOffset, localNameLength, centralStart)
      val localNameData = archive.copyOfRange(localNameOffset, localNameEnd)
      if (!localNameData.contentEquals(record.nameData)) {
        malformed("local and central paths differ for ${record.name}")
      }
      val dataEnd = checkedEnd(localNameEnd, record.compressedSize, centralStart)
      val payload = archive.copyOfRange(localNameEnd, dataEnd)
      payloads[record.name] = payload
      occupiedRanges += offset until dataEnd
    }

    var expectedOffset = 0
    occupiedRanges.sortedBy { it.first }.forEach { range ->
      if (range.first != expectedOffset || range.last + 1 < range.first) {
        malformed("overlapping entries, hidden bytes, or an archive preamble")
      }
      expectedOffset = range.last + 1
    }
    if (expectedOffset != centralStart) {
      malformed("hidden bytes before the central directory")
    }
    records.forEach { record ->
      val payload = payloads.getValue(record.name)
      if (crc32(payload) != record.crc32) {
        throw PiyoDeckImportException.CrcMismatch(record.name)
      }
    }
    return payloads
  }

  fun write(entries: List<PiyoDeckZipEntry>): ByteArray {
    val names = entries.map { it.name }
    if (names != listOf("manifest.json", "deck.json")) {
      throw PiyoDeckImportException.InvalidEntrySet(names)
    }
    val output = ByteAccumulator()
    val centralRecords = mutableListOf<WriterCentralRecord>()
    entries.forEach { entry ->
      val nameData = entry.name.toByteArray(StandardCharsets.US_ASCII)
      validateEntryName(entry.name, nameData)
      val localOffset = output.size
      val checksum = crc32(entry.data)
      output.uint32(LOCAL_FILE_HEADER_SIGNATURE)
      output.uint16(ZIP_VERSION)
      output.uint16(UTF8_FLAG)
      output.uint16(STORE_METHOD)
      output.uint16(DOS_TIME)
      output.uint16(DOS_DATE)
      output.uint32(checksum)
      output.uint32(entry.data.size.toLong())
      output.uint32(entry.data.size.toLong())
      output.uint16(nameData.size)
      output.uint16(0)
      output.bytes(nameData)
      output.bytes(entry.data)
      centralRecords += WriterCentralRecord(nameData, checksum, entry.data.size, localOffset)
    }

    val centralOffset = output.size
    centralRecords.forEach { record ->
      output.uint32(CENTRAL_DIRECTORY_SIGNATURE)
      output.uint16(ZIP_VERSION)
      output.uint16(ZIP_VERSION)
      output.uint16(UTF8_FLAG)
      output.uint16(STORE_METHOD)
      output.uint16(DOS_TIME)
      output.uint16(DOS_DATE)
      output.uint32(record.crc32)
      output.uint32(record.size.toLong())
      output.uint32(record.size.toLong())
      output.uint16(record.nameData.size)
      output.uint16(0)
      output.uint16(0)
      output.uint16(0)
      output.uint16(0)
      output.uint32(0)
      output.uint32(record.localHeaderOffset.toLong())
      output.bytes(record.nameData)
    }
    val centralSize = output.size - centralOffset
    output.uint32(END_OF_CENTRAL_DIRECTORY_SIGNATURE)
    output.uint16(0)
    output.uint16(0)
    output.uint16(entries.size)
    output.uint16(entries.size)
    output.uint32(centralSize.toLong())
    output.uint32(centralOffset.toLong())
    output.uint16(0)
    val packageData = output.toByteArray()
    if (packageData.size > PiyoDeckPackageLimits.MAXIMUM_PACKAGE_BYTES) {
      throw PiyoDeckImportException.PackageTooLarge(
        packageData.size,
        PiyoDeckPackageLimits.MAXIMUM_PACKAGE_BYTES,
      )
    }
    return packageData
  }

  private fun decodeEntryName(data: ByteArray): String {
    val name = try {
      StandardCharsets.UTF_8
        .newDecoder()
        .onMalformedInput(CodingErrorAction.REPORT)
        .onUnmappableCharacter(CodingErrorAction.REPORT)
        .decode(java.nio.ByteBuffer.wrap(data))
        .toString()
    } catch (error: Exception) {
      throw PiyoDeckImportException.UnsafeEntryPath("<invalid UTF-8>")
    }
    validateEntryName(name, data)
    return name
  }

  private fun validateEntryName(name: String, data: ByteArray) {
    if (data.any { it.toInt() and 0x80 != 0 } ||
      name.isEmpty() || name == "." || name == ".." ||
      '/' in name || '\\' in name || ':' in name || '\u0000' in name
    ) {
      throw PiyoDeckImportException.UnsafeEntryPath(name)
    }
    if (!name.toByteArray(StandardCharsets.US_ASCII).contentEquals(data)) {
      throw PiyoDeckImportException.UnsafeEntryPath(name)
    }
  }

  private fun ensureRange(start: Int, length: Int, limit: Int) {
    if (start < 0 || length < 0 || start > limit || length > limit - start) {
      malformed("record exceeds archive bounds")
    }
  }

  private fun checkedEnd(start: Int, length: Int, limit: Int): Int {
    ensureRange(start, length, limit)
    return start + length
  }

  private fun crc32(data: ByteArray): Long = CRC32().apply { update(data) }.value

  private fun malformed(reason: String): Nothing =
    throw PiyoDeckImportException.MalformedArchive(reason)

  private fun unsupported(feature: String): Nothing =
    throw PiyoDeckImportException.UnsupportedArchiveFeature(feature)

  private data class CentralRecord(
    val name: String,
    val nameData: ByteArray,
    val versionNeeded: Int,
    val flags: Int,
    val method: Int,
    val crc32: Long,
    val compressedSize: Int,
    val uncompressedSize: Int,
    val localHeaderOffset: Long,
  )

  private data class WriterCentralRecord(
    val nameData: ByteArray,
    val crc32: Long,
    val size: Int,
    val localHeaderOffset: Int,
  )
}

private fun ByteArray.uint16(offset: Int): Int {
  if (offset < 0 || offset > size - 2) {
    throw PiyoDeckImportException.MalformedArchive("truncated 16-bit field")
  }
  return (this[offset].toInt() and 0xFF) or ((this[offset + 1].toInt() and 0xFF) shl 8)
}

private fun ByteArray.uint32(offset: Int): Long {
  if (offset < 0 || offset > size - 4) {
    throw PiyoDeckImportException.MalformedArchive("truncated 32-bit field")
  }
  return (this[offset].toLong() and 0xFF) or
    ((this[offset + 1].toLong() and 0xFF) shl 8) or
    ((this[offset + 2].toLong() and 0xFF) shl 16) or
    ((this[offset + 3].toLong() and 0xFF) shl 24)
}

private class ByteAccumulator {
  private var bytes = ByteArray(1_024)
  var size: Int = 0
    private set

  fun uint16(value: Int) {
    require(value in 0..0xFFFF)
    ensureCapacity(2)
    bytes[size++] = value.toByte()
    bytes[size++] = (value ushr 8).toByte()
  }

  fun uint32(value: Long) {
    if (value !in 0..0xFFFF_FFFFL) {
      throw PiyoDeckImportException.UnsupportedArchiveFeature("ZIP64")
    }
    ensureCapacity(4)
    bytes[size++] = value.toByte()
    bytes[size++] = (value ushr 8).toByte()
    bytes[size++] = (value ushr 16).toByte()
    bytes[size++] = (value ushr 24).toByte()
  }

  fun bytes(value: ByteArray) {
    ensureCapacity(value.size)
    value.copyInto(bytes, destinationOffset = size)
    size += value.size
  }

  fun toByteArray(): ByteArray = bytes.copyOf(size)

  private fun ensureCapacity(additional: Int) {
    if (additional < 0 || size > Int.MAX_VALUE - additional) {
      throw PiyoDeckImportException.UnsupportedArchiveFeature("ZIP64")
    }
    val required = size + additional
    if (required <= bytes.size) return
    var capacity = bytes.size
    while (capacity < required) {
      val next = (capacity * 2L).coerceAtMost(Int.MAX_VALUE.toLong()).toInt()
      if (next <= capacity) throw PiyoDeckImportException.UnsupportedArchiveFeature("ZIP64")
      capacity = next
    }
    bytes = bytes.copyOf(capacity)
  }
}

File: ios/HangulEngine/Sources/DeckKit/PiyoDeckZIP.swift
import Foundation

struct PiyoDeckZIPEntry: Equatable, Sendable {
  let name: String
  let data: Data
}

enum PiyoDeckZIP {
  private static let localFileHeaderSignature: UInt32 = 0x0403_4B50
  private static let centralDirectoryHeaderSignature: UInt32 = 0x0201_4B50
  private static let endOfCentralDirectorySignature: UInt32 = 0x0605_4B50
  private static let utf8Flag: UInt16 = 0x0800
  private static let storeMethod: UInt16 = 0
  private static let minimumDOSDate: UInt16 = 0x0021
  private static let requiredEntryNames: Set<String> = ["manifest.json", "deck.json"]

  static func write(entries: [PiyoDeckZIPEntry]) throws -> Data {
    let names = entries.map(\.name)
    guard names == ["manifest.json", "deck.json"]
    else {
      throw PiyoDeckImportError.invalidEntrySet(names)
    }

    struct CentralRecord {
      let nameData: Data
      let crc32: UInt32
      let size: UInt32
      let localHeaderOffset: UInt32
    }

    var result = Data()
    var centralRecords: [CentralRecord] = []

    for entry in entries {
      guard isSafeEntryPath(entry.name), let nameData = entry.name.data(using: .utf8) else {
        throw PiyoDeckImportError.unsafeEntryPath(entry.name)
      }
      guard entry.data.count <= Int(UInt32.max), result.count <= Int(UInt32.max),
        nameData.count <= Int(UInt16.max)
      else {
        throw PiyoDeckImportError.unsupportedArchiveFeature("ZIP64")
      }

      let crc32 = PiyoDeckCRC32.checksum(entry.data)
      let size = UInt32(entry.data.count)
      let offset = UInt32(result.count)

      result.appendLittleEndian(localFileHeaderSignature)
      result.appendLittleEndian(UInt16(20))
      result.appendLittleEndian(utf8Flag)
      result.appendLittleEndian(storeMethod)
      result.appendLittleEndian(UInt16(0))
      result.appendLittleEndian(minimumDOSDate)
      result.appendLittleEndian(crc32)
      result.appendLittleEndian(size)
      result.appendLittleEndian(size)
      result.appendLittleEndian(UInt16(nameData.count))
      result.appendLittleEndian(UInt16(0))
      result.append(nameData)
      result.append(entry.data)

      centralRecords.append(
        CentralRecord(nameData: nameData, crc32: crc32, size: size, localHeaderOffset: offset)
      )
    }

    guard result.count <= Int(UInt32.max) else {
      throw PiyoDeckImportError.unsupportedArchiveFeature("ZIP64")
    }
    let centralDirectoryOffset = UInt32(result.count)

    for record in centralRecords {
      result.appendLittleEndian(centralDirectoryHeaderSignature)
      result.appendLittleEndian(UInt16(20))
      result.appendLittleEndian(UInt16(20))
      result.appendLittleEndian(utf8Flag)
      result.appendLittleEndian(storeMethod)
      result.appendLittleEndian(UInt16(0))
      result.appendLittleEndian(minimumDOSDate)
      result.appendLittleEndian(record.crc32)
      result.appendLittleEndian(record.size)
      result.appendLittleEndian(record.size)
      result.appendLittleEndian(UInt16(record.nameData.count))
      result.appendLittleEndian(UInt16(0))
      result.appendLittleEndian(UInt16(0))
      result.appendLittleEndian(UInt16(0))
      result.appendLittleEndian(UInt16(0))
      result.appendLittleEndian(UInt32(0))
      result.appendLittleEndian(record.localHeaderOffset)
      result.append(record.nameData)
    }

    let centralDirectorySize = result.count - Int(centralDirectoryOffset)
    guard centralDirectorySize <= Int(UInt32.max) else {
      throw PiyoDeckImportError.unsupportedArchiveFeature("ZIP64")
    }
    result.appendLittleEndian(endOfCentralDirectorySignature)
    result.appendLittleEndian(UInt16(0))
    result.appendLittleEndian(UInt16(0))
    result.appendLittleEndian(UInt16(centralRecords.count))
    result.appendLittleEndian(UInt16(centralRecords.count))
    result.appendLittleEndian(UInt32(centralDirectorySize))
    result.appendLittleEndian(centralDirectoryOffset)
    result.appendLittleEndian(UInt16(0))
    return result
  }

  static func read(_ archive: Data) throws -> [String: Data] {
    let endRecordSize = 22
    guard archive.count >= endRecordSize else {
      throw PiyoDeckImportError.malformedArchive("missing end-of-central-directory record")
    }
    let endOffset = archive.count - endRecordSize
    guard try archive.littleEndianUInt32(at: endOffset) == endOfCentralDirectorySignature else {
      throw PiyoDeckImportError.malformedArchive(
        "archive comments, trailing data, or a missing end record"
      )
    }

    let diskNumber = try archive.littleEndianUInt16(at: endOffset + 4)
    let centralDirectoryDisk = try archive.littleEndianUInt16(at: endOffset + 6)
    let entriesOnDisk = try archive.littleEndianUInt16(at: endOffset + 8)
    let totalEntries = try archive.littleEndianUInt16(at: endOffset + 10)
    let centralDirectorySize = Int(try archive.littleEndianUInt32(at: endOffset + 12))
    let centralDirectoryOffset = Int(try archive.littleEndianUInt32(at: endOffset + 16))
    let archiveCommentLength = try archive.littleEndianUInt16(at: endOffset + 20)

    guard diskNumber == 0, centralDirectoryDisk == 0 else {
      throw PiyoDeckImportError.unsupportedArchiveFeature("split archive")
    }
    guard archiveCommentLength == 0 else {
      throw PiyoDeckImportError.unsupportedArchiveFeature("archive comment")
    }
    guard entriesOnDisk == totalEntries else {
      throw PiyoDeckImportError.unsupportedArchiveFeature("multi-disk central directory")
    }
    guard totalEntries == 2 else {
      throw PiyoDeckImportError.invalidEntrySet([])
    }
    guard centralDirectoryOffset >= 0, centralDirectorySize >= 0,
      centralDirectoryOffset <= endOffset,
      centralDirectorySize == endOffset - centralDirectoryOffset
    else {
      throw PiyoDeckImportError.malformedArchive("invalid central-directory bounds")
    }

    struct CentralRecord {
      let name: String
      let nameData: Data
      let versionNeeded: UInt16
      let flags: UInt16
      let method: UInt16
      let crc32: UInt32
      let compressedSize: Int
      let uncompressedSize: Int
      let localHeaderOffset: Int
    }

    var cursor = centralDirectoryOffset
    var records: [CentralRecord] = []
    for _ in 0..<Int(totalEntries) {
      guard try archive.littleEndianUInt32(at: cursor) == centralDirectoryHeaderSignature else {
        throw PiyoDeckImportError.malformedArchive("invalid central-directory entry")
      }
      let versionMadeBy = try archive.littleEndianUInt16(at: cursor + 4)
      let versionNeeded = try archive.littleEndianUInt16(at: cursor + 6)
      let flags = try archive.littleEndianUInt16(at: cursor + 8)
      let method = try archive.littleEndianUInt16(at: cursor + 10)
      let crc32 = try archive.littleEndianUInt32(at: cursor + 16)
      let compressedSize32 = try archive.littleEndianUInt32(at: cursor + 20)
      let uncompressedSize32 = try archive.littleEndianUInt32(at: cursor + 24)
      let nameLength = Int(try archive.littleEndianUInt16(at: cursor + 28))
      let extraLength = Int(try archive.littleEndianUInt16(at: cursor + 30))
      let commentLength = Int(try archive.littleEndianUInt16(at: cursor + 32))
      let diskStart = try archive.littleEndianUInt16(at: cursor + 34)
      let externalAttributes = try archive.littleEndianUInt32(at: cursor + 38)
      let localHeaderOffset32 = try archive.littleEndianUInt32(at: cursor + 42)

      if versionNeeded > 20 {
        throw PiyoDeckImportError.unsupportedArchiveFeature("ZIP version \(versionNeeded)")
      }
      if compressedSize32 == UInt32.max || uncompressedSize32 == UInt32.max
        || localHeaderOffset32 == UInt32.max
      {
        throw PiyoDeckImportError.unsupportedArchiveFeature("ZIP64")
      }
      if flags & ~utf8Flag != 0 {
        let feature = flags & 0x0001 != 0 ? "encryption" : "general-purpose flag 0x\(hex(flags))"
        throw PiyoDeckImportError.unsupportedArchiveFeature(feature)
      }
      guard method == storeMethod else {
        throw PiyoDeckImportError.unsupportedArchiveFeature("compression method \(method)")
      }
      guard extraLength == 0 else {
        throw PiyoDeckImportError.unsupportedArchiveFeature("entry extra field")
      }
      guard commentLength == 0 else {
        throw PiyoDeckImportError.unsupportedArchiveFeature("entry comment")
      }
      guard diskStart == 0 else {
        throw PiyoDeckImportError.unsupportedArchiveFeature("split archive entry")
      }
      let hostSystem = UInt8(truncatingIfNeeded: versionMadeBy >> 8)
      let unixFileType = UInt16(truncatingIfNeeded: externalAttributes >> 16) & 0xF000
      if (hostSystem == 3 || hostSystem == 19) && unixFileType == 0xA000 {
        throw PiyoDeckImportError.unsupportedArchiveFeature("symbolic link")
      }
      if (hostSystem == 3 || hostSystem == 19) && unixFileType == 0x4000 {
        throw PiyoDeckImportError.unsupportedArchiveFeature("directory entry")
      }
      if (hostSystem == 3 || hostSystem == 19) && unixFileType != 0 && unixFileType != 0x8000 {
        throw PiyoDeckImportError.unsupportedArchiveFeature("non-regular entry")
      }
      if externalAttributes & 0x10 != 0 {
        throw PiyoDeckImportError.unsupportedArchiveFeature("directory entry")
      }

      let headerSize = 46
      let nameOffset = try checkedEnd(cursor, headerSize, limit: endOffset)
      let entryEnd = try checkedEnd(nameOffset, nameLength, limit: endOffset)
      let nameData = archive.subdata(in: nameOffset..<entryEnd)
      guard let name = String(data: nameData, encoding: .utf8), name.data(using: .utf8) == nameData
      else {
        throw PiyoDeckImportError.unsafeEntryPath("<invalid UTF-8>")
      }
      guard isSafeEntryPath(name) else {
        throw PiyoDeckImportError.unsafeEntryPath(name)
      }
      let compressedSize = Int(compressedSize32)
      let uncompressedSize = Int(uncompressedSize32)
      guard compressedSize == uncompressedSize else {
        throw PiyoDeckImportError.malformedArchive("STORE entry sizes differ for \(name)")
      }
      records.append(
        CentralRecord(
          name: name,
          nameData: nameData,
          versionNeeded: versionNeeded,
          flags: flags,
          method: method,
          crc32: crc32,
          compressedSize: compressedSize,
          uncompressedSize: uncompressedSize,
          localHeaderOffset: Int(localHeaderOffset32)
        )
      )
      cursor = entryEnd
    }
    guard cursor == endOffset else {
      throw PiyoDeckImportError.malformedArchive("unused central-directory bytes")
    }

    let names = records.map(\.name)
    guard Set(names) == requiredEntryNames, Set(names).count == names.count else {
      throw PiyoDeckImportError.invalidEntrySet(names)
    }

    var payloads: [String: Data] = [:]
    var occupiedRanges: [(start: Int, end: Int)] = []
    for record in records {
      let maximum =
        record.name == "manifest.json"
        ? PiyoDeckPackageLimits.maximumManifestBytes
        : PiyoDeckPackageLimits.maximumDeckBytes
      guard record.uncompressedSize <= maximum else {
        throw PiyoDeckImportError.entryTooLarge(
          name: record.name,
          actual: record.uncompressedSize,
          maximum: maximum
        )
      }
      let offset = record.localHeaderOffset
      guard try archive.littleEndianUInt32(at: offset) == localFileHeaderSignature else {
        throw PiyoDeckImportError.malformedArchive("missing local header for \(record.name)")
      }
      let localVersionNeeded = try archive.littleEndianUInt16(at: offset + 4)
      let localFlags = try archive.littleEndianUInt16(at: offset + 6)
      let localMethod = try archive.littleEndianUInt16(at: offset + 8)
      let localCRC32 = try archive.littleEndianUInt32(at: offset + 14)
      let localCompressedSize = Int(try archive.littleEndianUInt32(at: offset + 18))
      let localUncompressedSize = Int(try archive.littleEndianUInt32(at: offset + 22))
      let localNameLength = Int(try archive.littleEndianUInt16(at: offset + 26))
      let localExtraLength = Int(try archive.littleEndianUInt16(at: offset + 28))

      guard localVersionNeeded == record.versionNeeded, localFlags == record.flags,
        localMethod == record.method, localCRC32 == record.crc32,
        localCompressedSize == record.compressedSize,
        localUncompressedSize == record.uncompressedSize
      else {
        throw PiyoDeckImportError.malformedArchive(
          "local and central headers differ for \(record.name)"
        )
      }
      guard localExtraLength == 0 else {
        throw PiyoDeckImportError.unsupportedArchiveFeature("local entry extra field")
      }
      let localNameOffset = try checkedEnd(offset, 30, limit: centralDirectoryOffset)
      let localNameEnd = try checkedEnd(
        localNameOffset, localNameLength, limit: centralDirectoryOffset)
      let localNameData = archive.subdata(in: localNameOffset..<localNameEnd)
      guard localNameData == record.nameData else {
        throw PiyoDeckImportError.malformedArchive(
          "local and central paths differ for \(record.name)"
        )
      }
      let dataEnd = try checkedEnd(
        localNameEnd, record.compressedSize, limit: centralDirectoryOffset)
      let payload = archive.subdata(in: localNameEnd..<dataEnd)
      payloads[record.name] = payload
      occupiedRanges.append((start: offset, end: dataEnd))
    }

    let sortedRanges = occupiedRanges.sorted { $0.start < $1.start }
    var expectedOffset = 0
    for range in sortedRanges {
      guard range.start == expectedOffset, range.end >= range.start else {
        throw PiyoDeckImportError.malformedArchive(
          "overlapping entries, hidden bytes, or an archive preamble"
        )
      }
      expectedOffset = range.end
    }
    guard expectedOffset == centralDirectoryOffset else {
      throw PiyoDeckImportError.malformedArchive("hidden bytes before the central directory")
    }
    for record in records {
      guard let payload = payloads[record.name], PiyoDeckCRC32.checksum(payload) == record.crc32
      else {
        throw PiyoDeckImportError.crcMismatch(name: record.name)
      }
    }
    return payloads
  }

  private static func checkedEnd(_ start: Int, _ length: Int, limit: Int) throws -> Int {
    guard start >= 0, length >= 0, start <= limit, length <= limit - start else {
      throw PiyoDeckImportError.malformedArchive("record exceeds archive bounds")
    }
    return start + length
  }

  private static func isSafeEntryPath(_ name: String) -> Bool {
    guard !name.isEmpty, name.unicodeScalars.allSatisfy({ $0.value < 0x80 }) else { return false }
    guard !name.contains("/"), !name.contains("\\"), !name.contains(":"), !name.contains("\0")
    else { return false }
    return name != "." && name != ".."
  }

  private static func hex(_ value: UInt16) -> String {
    String(value, radix: 16, uppercase: false)
  }
}

enum PiyoDeckCRC32 {
  private static let table: [UInt32] = (0..<256).map { index in
    var value = UInt32(index)
    for _ in 0..<8 {
      value = value & 1 == 1 ? 0xEDB8_8320 ^ (value >> 1) : value >> 1
    }
    return value
  }

  static func checksum(_ data: Data) -> UInt32 {
    var crc: UInt32 = 0xFFFF_FFFF
    for byte in data {
      let index = Int((crc ^ UInt32(byte)) & 0xFF)
      crc = table[index] ^ (crc >> 8)
    }
    return crc ^ 0xFFFF_FFFF
  }
}

extension Data {
  fileprivate mutating func appendLittleEndian(_ value: UInt16) {
    append(UInt8(truncatingIfNeeded: value))
    append(UInt8(truncatingIfNeeded: value >> 8))
  }

  fileprivate mutating func appendLittleEndian(_ value: UInt32) {
    append(UInt8(truncatingIfNeeded: value))
    append(UInt8(truncatingIfNeeded: value >> 8))
    append(UInt8(truncatingIfNeeded: value >> 16))
    append(UInt8(truncatingIfNeeded: value >> 24))
  }

  fileprivate func littleEndianUInt16(at offset: Int) throws -> UInt16 {
    guard offset >= 0, offset <= count - 2 else {
      throw PiyoDeckImportError.malformedArchive("truncated 16-bit field")
    }
    let index = startIndex + offset
    return UInt16(self[index]) | UInt16(self[index + 1]) << 8
  }

  fileprivate func littleEndianUInt32(at offset: Int) throws -> UInt32 {
    guard offset >= 0, offset <= count - 4 else {
      throw PiyoDeckImportError.malformedArchive("truncated 32-bit field")
    }
    let index = startIndex + offset
    return UInt32(self[index]) | UInt32(self[index + 1]) << 8
      | UInt32(self[index + 2]) << 16 | UInt32(self[index + 3]) << 24
  }
}
